Add event support to SC3ML 0.10 schema

Brtle opened the pull request:
Hey! I looked a bit to the SC3ML 0.10 event support (see https://github.com/obspy/obspy/issues/2004) and it should be pretty good. There are some changes on the `Arrival/weight` mapping and some fields have been added. Event types have also been added to better match QuakeML. The rest concerns the inventory part that I didn't touch.

I made some changes to the XSLT files provided by SeisComP3 and as they added them to their [Github repository](https://github.com/SeisComP3/seiscomp3/tree/master/src/trunk/libs/xml), I did a [pull request](https://github.com/SeisComP3/seiscomp3/pull/158). It is still open so changes can still be made to this PR.

Morevover, the SC3ML files are always written in SC3ML 0.10. I didn't let the possibility to write files in SC3ML 0.9. What do you think about that? Is it worth to keep the choice of the version at the cost of a slightly greater complexity? I do not know if many people want to write in an old SC3ML version.

megies commented on obspy/io/seiscomp/core.py:
```diff
@@ -87,6 +87,9 @@ def validate(path_or_object, version='0.9', verbose=False):
     :return: `True` if SC3ML file is valid.
     """
+    if version is None:
+        version = SUPPORTED_XSD_VERSION[-1]
```
Could we not also look up the version from the file that is to be validated?

Brtle replied:
You're right! I changed that.

megies commented on obspy/io/seiscomp/event.py:
```diff
@@ -116,7 +116,7 @@ def _write_sc3ml(catalog, filename, validate=False, verbose=False,
     quakeml_doc = Pickler(nsmap=nsmap_).dumps(catalog)
     xslt_filename = os.path.join(os.path.dirname(__file__), 'data',
-                                 'quakeml_1.2__sc3ml_0.9.xsl')
+                                 'quakeml_1.2__sc3ml_0.10.xsl')
```
No idea.. do we want to be able to write different versions of SC3ML? Probably not and writing the most recent version is OK to keep maintenance down, I guess..

megies commented:
Looks like you're changing all of the test files to SC3ML 0.10 in their header.. I think we might want to keep at least some of them at 0.9 so that reading that older version stays tested?

Brtle commented:
I think most of the unit tests should be done on the latest version. Note that all XSLT files are exactly the same, except the number version. However, it's true that there is a lack of tests for older versions. I added some unit tests to check the [qml-example-1.2-RC3.sc3ml](https://github.com/obspy/obspy/blob/master/obspy/io/seiscomp/tests/data/qml-example-1.2-RC3.sc3ml) file with older versions. To avoid adding test files (which are already numerous) I modified the version dynamically for each test. Let me know what you think about it and if the old versions are enough tested.

krischer commented:
Sorry the slow response time - this looks pretty reasonable to me. Is this ready - then I'd give it a more detailed review.

Brtle commented:
You can take a look! There are 2 things which can eventually be improved. The new SC3ML files are always written in SC3ML 0.10. Do you think we need to add the possibility to write 0.9 files? Moreover, I updated all the unit tests with the 0.10 schema. I added some unit tests to check old versions but it may not be enough. What do you think about it?

krischer commented:
> You can take a look! There are 2 things which can eventually be improved. The new SC3ML files are always written in SC3ML 0.10. Do you think we need to add the possibility to write 0.9 files? Moreover, I updated all the unit tests with the 0.10 schema. I added some unit tests to check old versions but it may not be enough. What do you think about it?

In the tests for reading the old versions you basically already write old SC3ML files. How about exposing that functionality to the `_write_sc3ml()` function and adding a `version` argument. Then you can also use this for the same tests and already have some reasonable coverage. All xsd schemas for all SC3ML versions are also already present and if you also use these in the tests I would say we could have quite some confidence that SC3ML writing works correctly. All of this seems to pretty easy to do with the XSLT approach so I'd say go for it!



The new fields in SC3ML 0.10 (I guess its mainly the `pdf` fields) could be to custom tags in a separate namespace (see e.g. here for an example: http://docs.obspy.org/tutorial/code_snippets/quakeml_custom_tags.html). At that point ObsPy could directly read/write it. This would require some modification to the XSLT files and I don't have enough experience with these files to judge the effort.

Brtle commented:
> In the tests for reading the old versions you basically already write old SC3ML files. How about exposing that functionality to the _write_sc3ml() function and adding a version argument. Then you can also use this for the same tests and already have some reasonable coverage. All xsd schemas for all SC3ML versions are also already present and if you also use these in the tests I would say we could have quite some confidence that SC3ML writing works correctly. All of this seems to pretty easy to do with the XSLT approach so I'd say go for it!

The point is that I juste change the namespace. The [qml-example-1.2-RC3.sc3ml](https://github.com/obspy/obspy/blob/master/obspy/io/seiscomp/tests/data/qml-example-1.2-RC3.sc3ml) is actually the same for every SC3ML version, except the namespace part. I just wanted to test that old versions conversion didn't crash without testing every distinctive feature of each version. If we add the possibility to change the namespace in the `_write_sc3ml` function, this can generate unvalid SC3ML files. This can be done but a warning message should be displayed.

What we can do is maybe test all the versions that a unit test can run, for instance:

```python
def test_read_xslt_origin(self, version=['0.9', '0.10']):
    …

def test_read_xslt_magnitude(self, version=['0.8', '0.9', '0.10']):
    …
```

So old versions should be relatively well tested without exploding the number of unit tests.

> The new fields in SC3ML 0.10 (I guess its mainly the pdf fields) could be to custom tags in a separate namespace (see e.g. here for an example: http://docs.obspy.org/tutorial/code_snippets/quakeml_custom_tags.html). At that point ObsPy could directly read/write it. This would require some modification to the XSLT files and I don't have enough experience with these files to judge the effort.

The namespace is hardcoded in the XSLT files so we can't use the same XSLT file without templating the file itself. The easiest solution seems to keep one file per version. However, we will need to maintain several XSLT files when most people will likely use the latest version. It's easier to maintain in the read plugin since only the namespace change between all sc3ml_0.*__quakeml_1.2.xsl files.

Brtle commented:
Sorry, it's been a little too long since I did not look into this PR!

I did some modifications to the unit tests. For each read conversion tests, every SC3ML versions for which the file is valid are tested. So old versions are not completly tested but I think this is the best compromise we can do.

I didn't change the `_write_sc3ml` method. I think it can be misleading if we add a parameter `version` which only change the namespace as I did in the unit tests. Invalid SC3ML files could be generated.

> The new fields in SC3ML 0.10 (I guess its mainly the pdf fields) could be to custom tags in a separate namespace (see e.g. here for an example: http://docs.obspy.org/tutorial/code_snippets/quakeml_custom_tags.html). At that point ObsPy could directly read/write it. This would require some modification to the XSLT files and I don't have enough experience with these files to judge the effort.

I had answered a little quickly in my last reply and I did not quite understand it. But you're right, these custom tags should be used during the conversion! When I made the plugin the first time I was not aware of that. Actually there are [other nodes](https://github.com/obspy/obspy/blob/master/obspy/io/seiscomp/data/sc3ml_0.9__quakeml_1.2.xsl#L92) for which it could be used. However, this require some upstream modifications to the XSLT files. Since this doesn't seem completely necessary and it does not only concern the 0.10 schema, it would be perhaps more judicious to make another PR?

megies commented:
> and it does not only concern the 0.10 schema, it would be perhaps more judicious to make another PR?

yeah, sounds like it can be done in a separate PR if you prefer not to work on it right now

megies commented:
This looks ready @Brtle? Anything left to do in here?

krischer commented:
Thanks for this! I'll add a changelog entry directly to the master branch.
